File: QytetetRuby/lib/vista_textual_qytetet.rb
#///////////////////////////////////////////////////////////////////////////////
#// Autor: Alejandro Manzanares Lemus
#// 2º GII A1
#///////////////////////////////////////////////////////////////////////////////
require_relative "qytetet.rb"
require_relative "controlador_qytetet.rb"
require_relative "opcion_menu.rb"
module ModuloVistaTextualQytetet
  
  class VistaTextualQytetet
#    include Qytetet
#    include ControladorQytetet
    
    @@controlador = ModuloControladorQytetet::ControladorQytetet.instance
    
    def initialize
      
    end
    
    def obtener_nombre_jugadores()
      puts "Introduce el numero de jugadores: "
     numero_jugadores_valido = Array.new
      for i in 2..ModeloQytetet::Qytetet.get_max_jugadores()
          numero_jugadores_valido<<(i.to_s);
      end

      num_jugadores = leer_valor_correcto(numero_jugadores_valido).to_i()

      nombres = Array.new
      for i in 0..num_jugadores-1
          puts "Introduce el nombre del jugador: "
          nombres<<(gets.chomp())
      end
      nombres
    end
    
    def elegir_operacion()
      aux = @@controlador.obtener_operaciones_juego_validas()
      if aux.empty?
          return (-1)
      else
          aux_str = Array.new
          for numero in aux
              aux_str<<(numero.to_s())
          end
          leer_valor_correcto(aux_str).to_i()
      end
    end
    
    def elegir_casilla(opcion_menu)
      aux = @@controlador.obtener_casillas_validas(opcion_menu);
      if aux.empty?
          return (-1)
      else
          aux_str = Array.new
          for numero in aux
              aux_str<<(numero.to_s())
          end
          leer_valor_correcto(aux_str).to_i()
      end
    end
    
    def leer_valor_correcto(valores_correctos)
      entrada = ""
      loop do
        entrada = gets.chomp
        break if valores_correctos.include?(entrada)
        end
      entrada
    end
    
    def imprimir_operaciones_juego_validas()
        aux = @@controlador.obtener_operaciones_juego_validas()
        for i in aux
            puts "#{i}  #{ModuloControladorQytetet::OpcionMenu[i]}"
        end
    end
    
    def self.main()
      ui = VistaTextualQytetet.new
      @@controlador.nombre_jugadores = ui.obtener_nombre_jugadores()
      operacion_elegida = 0
      casilla_elegida = 0

      loop do
        ui.imprimir_operaciones_juego_validas()
        operacion_elegida = ui.elegir_operacion()
        necesita_elegir_casilla = @@controlador.necesita_elegir_casilla(operacion_elegida);
        if necesita_elegir_casilla
            puts "Por favor, elige la casilla:\n"
            casilla_elegida = ui.elegir_casilla(operacion_elegida)
        end
        if !necesita_elegir_casilla || casilla_elegida >= 0
          puts @@controlador.realizar_operacion(operacion_elegida, casilla_elegida)
        end
      end
    end
  end
  VistaTextualQytetet.main
end

File: QytetetJava/src/vistatextualqytetet/VistaTextualQytetet.java
////////////////////////////////////////////////////////////////////////////////
// Autor: Alejandro Manzanares Lemus
// 2º GII A1
////////////////////////////////////////////////////////////////////////////////
package vistatextualqytetet;

import controladorqytetet.*;
import java.util.Scanner;
import java.util.ArrayList;
import modeloqytetet.Qytetet;

        
public class VistaTextualQytetet {
    ControladorQytetet controlador  = ControladorQytetet.getInstance();
    private static final Scanner in = new Scanner (System.in);

    public VistaTextualQytetet() {
    }
    
    public ArrayList<String> obtenerNombreJugadores(){
        System.out.println("Introduce el numero de jugadores: ");
        ArrayList<String> numero_jugadores_valido = new ArrayList<>();
        for(int i = 2; i <= Qytetet.MAX_JUGADORES; i++){
            numero_jugadores_valido.add(Integer.toString(i));
        }
        
        int num_jugadores = Integer.valueOf(leerValorCorrecto(numero_jugadores_valido));
        
        ArrayList<String> nombres = new ArrayList<>();
        for(int i = 0; i < num_jugadores; i++){
            System.out.println("Introduce el nombre del jugador: ");
            nombres.add(in.nextLine());
        }
        return nombres;
    }   
    public Integer elegirOperacion(){
        ArrayList<Integer> aux = controlador.obtenerOperacionesJuegoValidas();
        ArrayList<String> aux_str = new ArrayList<>();
        for(Integer numero: aux){
            aux_str.add(Integer.toString(numero));
        }
        
        return Integer.valueOf(leerValorCorrecto(aux_str));
    }
    public Integer elegirCasilla(int opcionMenu){
        ArrayList<Integer> aux;
        aux = controlador.obtenerCasillasValidas(opcionMenu);
        if(aux.isEmpty()){
            return -1;
        }
        else{
            ArrayList<String> aux_str = new ArrayList<String>();
            for(Integer numero: aux){
                aux_str.add(Integer.toString(numero));
            }
            return Integer.valueOf(leerValorCorrecto(aux_str));
        }
    }
    public String leerValorCorrecto(ArrayList<String> valoresCorrectos){
        String entrada;
        do{
            entrada = in.nextLine();
        }while(!valoresCorrectos.contains(entrada));
        
        return entrada;
    }    
    
    public void imprimirOperacionesJuegoValidas(){
        ArrayList<Integer> aux = controlador.obtenerOperacionesJuegoValidas();
        for(int i: aux){
            System.out.print(i);
            System.out.print(" ");
            System.out.println(OpcionMenu.values()[i]);
        }
    }
    
    public static void main(String args[]){
        VistaTextualQytetet ui = new VistaTextualQytetet();
        ui.controlador.setNombreJugadores(ui.obtenerNombreJugadores());
        int operacionElegida, casillaElegida = 0;
        boolean necesitaElegirCasilla;
        
        do {
            ui.imprimirOperacionesJuegoValidas();
            operacionElegida = ui.elegirOperacion();
            necesitaElegirCasilla = ui.controlador.necesitaElegirCasilla(operacionElegida);
            if (necesitaElegirCasilla){
                System.out.println("Por favor, elige la casilla:\n");
                casillaElegida = ui.elegirCasilla(operacionElegida);
            }
            if (!necesitaElegirCasilla || casillaElegida >= 0)
            System.out.println(ui.controlador.realizarOperacion(operacionElegida,
            casillaElegida));
        } while(true);
    }
}


